### ISSUE TITLE
Race condition for http connection in remote_storage.py

### ISSUE THREAD

**Issue by cout** [graphite-project/graphite-web#1353]
Running latest master, making numerous concurrent requests for metrics which hit the same remote, I am seeing a lot of exceptions in the log like this:

    Exception: Failed after 0 retry! See: Passive remote fetch failed to find cached results
    Thu Oct 15 18:04:45 2015 :: Error requesting http://10.203.11.21:8888/render/?target=%2A.me1-pre.interface-%2A.if_errors.tx&format=pickle&local=1&noCache=1&from=1444921484&until=1444932284
    Traceback (most recent call last):
      File "./webapp/graphite/remote_storage.py", line 182, in wait_for_results
         response = self.connection.getresponse()
     AttributeError: 'NoneType' object has no attribute 'getresponse'

I believe this is the result of a race condition:
* Thread 1 acquires the request lock
* Thread 1 blocks attempting to establish an http connection
* Thread 2 attempts to acquire the request lock but is unable to
* Thread 2 returns a FetchInProgress
* Thread 2 via the FetchInProgress object later calls wait_for_results()
* Thread 2 attempts to call self.connection.getresponse() on self.connection which is None
* Thread 1 successfully creates the http connection and assigns to self.connection

<del>
The effect can be mitigated by spinning to wait for self.connection to be assigned:

    179     def wait_for_results():                                           
    180       if wait_lock.acquire(False): # the FetchInProgress that gets waited on waits for the actual completion
    181         try:
    182           # spin waiting on connection        
    183           while not self.connection:
    184             pass        
    185           response = self.connection.getresponse()

But this has the undesirable effect of the waiting thread never completing if the requesting thread raises an exception.
</del>

Increasing MAX_FETCH_RETRIES may also prevent the effect from being visible to the user, at the expense of extraneous logging.

This issue may be a duplicate of #987.

EDIT: the spinning on the connection mentioned above does not work; I misunderstood how the RemoteReader class works.  I believe a recursive lock is the simplest solution (see comment below).

**Comment by cout** [comment]
I believe I have a solution.  Inside wait_for_results(), I think it is expected that the first thread to acquire request_lock is also the first thread to acquire wait_lock.

Changing request_lock to be an RLock instead of a Lock allows wait_for_results() to use request_lock instead of wait_lock (and wait_lock can be removed).  This ensures that the thread which initiated the http request is also the one which checks for the response; the remaining thread wait on the completion event.

**Comment by redbaron** [comment]
@cout , @wooparadog thanks for researching and providing a patch. I tested it with my current setup and seems to be working fine now.

**Comment by deniszh** [comment]
@cout @wooparadog @redbaron - does someone want to make a PR? Or I can do that.

**Comment by redbaron** [comment]
new feedback: it solved exceptions mentioned in the report, but installing patch on just 1 server crippled whole cluster  when memcache TTL start to expire. 

`wait_for_results` started to fail with socket timeouts massively on the server with patch (rest were left unpatched):

```
  File "/opt/graphite-web/webapp/graphite/remote_storage.py", line 185, in wait_for_results
    response = self.connection.getresponse()
  File "/opt/rh/python27/root/usr/lib64/python2.7/httplib.py", line 1073, in getresponse
    response.begin()
  File "/opt/rh/python27/root/usr/lib64/python2.7/httplib.py", line 415, in begin
    version, status, reason = self._read_status()
  File "/opt/rh/python27/root/usr/lib64/python2.7/httplib.py", line 371, in _read_status
    line = self.fp.readline(_MAXLINE + 1)
  File "/opt/rh/python27/root/usr/lib64/python2.7/socket.py", line 476, in readline
    data = self._sock.recv(self._rbufsize)
```

**Comment by redbaron** [comment]
I still dont like how `RemoteReader.request_cache` is used, it can potentially return really stale results server had enough workers and `settings.REMOTE_READER_CACHE_SIZE_LIMIT` is not reached for a long time

**Comment by redbaron** [comment]
Run it overnight, even though it fixed problem reported in this issue,there are still occasinional failures of passive writers without corresponding failure in the main fetcher:

```
Thu Jun 30 08:50:32 2016 :: Failed to complete subfetch
Traceback (most recent call last):
  File "/opt/graphite-web/webapp/graphite/readers.py", line 52, in fetch
    results[i] = result.waitForResults()
  File "/opt/graphite-web/webapp/graphite/readers.py", line 29, in waitForResults
    return self.wait_callback()
  File "/opt/graphite-web/webapp/graphite/remote_storage.py", line 219, in extract_my_results
    for series in wait_for_results():
  File "/opt/graphite-web/webapp/graphite/remote_storage.py", line 216, in wait_for_results
    raise Exception("Passive remote fetch failed to find cached results")
Exception: Passive remote fetch failed to find cached results
```


I run *single threaded processes* (mod_wsgi threads=1) so it is not a race between threads. Only way I see it to be happening is that  *within same thread*  `RemoteReader.fetch` are called not in the same order as corresponding `FetchInProgress.waitForResult` , this would cause passive checks to fail with no error from  `RemoteReader` which initiated a request.

Looking at the  code I am puzzled how it can happen:

```python
class MultiReader:
....
  def fetch(self, startTime, endTime):
    # Start the fetch on each node
    results = [ n.fetch(startTime, endTime) for n in self.nodes ]

    # Wait for any asynchronous operations to complete
    for i, result in enumerate(results):
      if isinstance(result, FetchInProgress):
        try:
          results[i] = result.waitForResults()
        except:
          log.exception("Failed to complete subfetch")
          results[i] = None
...
```

any ideas?
